### ISSUE TITLE
Tensor parallelism accuracy + grad norm scale issues

### ISSUE THREAD

**Issue by nathan-az** [pytorch/torchtune#2641]
I noticed that when comparing TP to DP with the same seed and adjusting gradient accumulation steps, there's an odd spike after the second training step.

<img width="389" alt="Image" src="https://github.com/user-attachments/assets/778d4bba-300d-4f51-844a-08c96da5eac6" />

I made a custom recipe with some additional metrics - logging the number of tokens per update step, number of correctly predicted tokens (top 1 accuracy from the logits), and accuracy. I saw a steep drop-off in accuracy at the same point.

<img width="1178" alt="Image" src="https://github.com/user-attachments/assets/445e8be9-aca2-4baa-a6ea-71d2ce75a0fb" />

I did some further testing (on cheaper hardware, hence smaller sizes), comparing two custom recipes:

1. tweak of `full_finetune_distributed` which includes the accuracy metrics. Metrics groups prefixed `acc`
2. tweak of a recipe following the stepwise pattern from https://github.com/pytorch/torchtune/pull/2543. Metrics groups prefixed `stepwise`

<img width="1076" alt="Image" src="https://github.com/user-attachments/assets/c8a574bc-2751-4976-82fb-b75f086b75cd" />

(2) is really just a slight re-formulation of the training loop which scales losses rather than using the gradient scaling. Note that although the scales of the losses are different between the original method and the step-wise method, for stepwise TP, stepwise DP, and original TP, accuracies are almost identical, while original TP is near zero.

This might indicate there is an issue with the gradient scaling. Although given quirks with my hardware I'd love if anyone could try to reproduce the results.

DP config:

<details>

```yaml
batch_size: 1
checkpointer:
  _component_: torchtune.training.FullModelHFCheckpointer
  checkpoint_files:
    filename_format: model-{}-of-{}.safetensors
    max_filename: '00030'
  model_type: LLAMA3
  output_dir: ${output_dir}
  recipe_checkpoint: null
  checkpoint_dir: inputs/model
clip_grad_norm: null
compile: true
custom_sharded_layers: []

data_parallel_shard_dim: 8
data_parallel_replicate_dim: 1
tensor_parallel_dim: 1

tensor_parallel_plan:
  _component_: torchtune.models.llama3.base_llama_tp_plan

dataset:
  _component_: torchtune.datasets.slimorca_dataset
  packed: true
  train_on_input: true

device: cuda
dtype: bf16
enable_activation_checkpointing: true
enable_activation_offloading: false
epochs: 1
fsdp_cpu_offload: false
gradient_accumulation_steps: 2
log_every_n_steps: 1
log_peak_memory_stats: true
loss:
  _component_: torch.nn.CrossEntropyLoss
  reduction: mean
max_steps_per_epoch: 20
metric_logger:
  _component_: torchtune.training.metric_logging.MLFlowLogger
optimizer:
  _component_: torchao.optim.AdamW8bit
  lr: 4.0e-06
optimizer_in_bwd: false
output_dir: outputs
resume_from_checkpoint: false
seed: 100
shuffle: true
tokenizer:
  max_seq_len: 4096
  path: inputs/model/original/tokenizer.model
  _component_: torchtune.models.llama3.llama3_tokenizer
model:
  _component_: torchtune.models.llama3_1.llama3_1_8b
```

</details>

The TP config is the above, with `tensor_parallel_dim: 8` (or your GPU count) and `gradient_accumulation_steps: 16` (or 2x your GPU count). Any help testing to see if this can be reproduced would be great, or any clear issues with my config. 

My setup used torch stable 2.7.0, torchao nightly, all built for CUDA 11.8. Some testing was done on H100s, some on L40S.

**Comment by joecummings** [comment]
I'll try to reproduce today - I have access to A100 and H100.

**Comment by joecummings** [comment]
Alright so I looked into this. 

To start, with your config and simply toggling between DP 8 and TP 8, I *am* able to reproduce the above loss spike. However, once I properly scale the batch size by 8 for TP, the loss spike goes away and the model appears to train as expected. This should be the first thing you try.

<img width="664" alt="Image" src="https://github.com/user-attachments/assets/6648b5d6-cc21-4e75-8854-08a4348280d6" />

So for DP with a batch size of 1, gradient accumulation of 16 steps, and 8 machines the effective batch size is 128. For TP with batch size of 1, gradient accumulation of 16 steps, and 8 machines, the effective batch size is just 16. 

A smaller batch size in and of itself does not cause instability, but with a low bit optimizer and unchanged learning rate? That's where problems come in. When I try with regular AdamW, I *do not* see the loss spike (although losses will not match b/c there's different data at each step without adjusting batch sizes).

<img width="664" alt="Image" src="https://github.com/user-attachments/assets/3dc98f40-ea0f-428e-ad61-d08f7d0a5c92" />

Similarly, if I scale the learning rate way down (4e-08) with the low bit optimizer, the initial loss spike also appears to goes away.

<img width="668" alt="Image" src="https://github.com/user-attachments/assets/cf7380eb-1f46-4557-b436-657447a062f5" />

**Comment by nathan-az** [comment]
Thanks @joecummings. Perhaps I wasn't clear on one detail: the configurations I used were:
* TP8 with bs=1, GA=16, and
* DP8 with bs=1, GA=2

So the effective batch sizes should have been the same. I just scale gradient accumulation rather than batch size to avoid having to worry about OOM. Are you suggesting that using the low bit optimiser could be the issue?

**Comment by joecummings** [comment]
Ahh! Sorry, I didn't understand that.

Okay, I reran the experiments and here are my thoughts.

1. With regular AdamW, things work as expected suggesting that some of the instability comes from the low bit optimizer.
2. Scaling via batch size, rather than gradient accumulation *also* seems to work better which suggests there might be something weird going on in our gradient accumulation logic. This warrants further investigation.

<img width="671" alt="Image" src="https://github.com/user-attachments/assets/f646594b-6e57-4299-88a3-204a90eb4acc" />

**Comment by ebsmothers** [comment]
@nathan-az just to confirm, is the scale_grads update sufficient to resolve this? If so let's close this one so we can focus on some of the remaining issues

**Comment by nathan-az** [comment]
@ebsmothers I believe so. I'll close for now, with two notes:

* Because the fix isn't obvious and was largely results-driven on my end, I'd love for somebody to repro it (should be as simple as running the same config on commits before/after the fix and seeing whether one is unstable and the other is stable)
* We should look to revert aspects of the fix PR in future, when:
    * `foreach` is supported for DTensor (if this is possible) since we explicitly disable `foreach` with TP
    * the actual in-place `.mul_` works for DTensor, since the fix might cause memory to peak for larger grads like `output` or `tok_embeddings`
